### PR TITLE
Introduce an abstract super class for aggregators.

### PR DESCRIPTION
**Pull request by jkwatson**
This is to hold the boilerplate logic and get it out of the specific implementations.

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics.aggregator;
+
+public abstract class AbstractAggregator implements Aggregator {
+
+  @Override
+  public void mergeToAndReset(Aggregator other) {
+    if (!other.getClass().isAssignableFrom(this.getClass())) {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: I heard once that getClass is expensive, maybe we do cache that in a constant for this.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * Merges the current value into the given {@code aggregator} and resets the current value in this
+   * {@code Aggregator}.
+   *
+   * <p>If this method is called, you can assume that the passed in aggregator can be cast to your
+   * self-type.
+   *
+   * @param aggregator The aggregator to merge with.
+   */
+  public abstract void doMergeAndReset(Aggregator aggregator);
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: Does this need to be public?

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics.aggregator;
+
+public abstract class AbstractAggregator implements Aggregator {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: Does this need to be public?

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 1 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: I am probably wrong, we need benchmarks :) later maybe

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: heh. yeah, I think getClass() is very fast, but benchmarks can show the truth.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 2 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: it does not. will fix.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 3 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: ditto

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 2 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: fixed

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 3 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: fixed

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java (hunk 1 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractAggregator.java]: also, I had this logic backwards. fixed now.

**Pull request merged**